### ISSUE TITLE
Error: NJS-024: memory allocation failed

### ISSUE THREAD

**Issue by souzacristsf** [oracle/node-oracledb#772]
Good Morning
I'm having a great project using this lib .. But at the beginning of the project creation I already faced this problem.

Error: NJS-024: memory allocation failed

practically this query brings me a quantity of ten thousand lines.

could anyone help me how to solve it?
Or if you have another alternative lib to use.

att,
Michel Ferreira Souza

**Comment by dmcghan** [comment]
Have you looked into the [query stream API](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#-913-streaming-query-results)?

**Comment by cjbj** [comment]
Or even the Result Set API: https://github.com/oracle/node-oracledb/blob/master/doc/api.md#resultsethandling

**Comment by dmcghan** [comment]
Sorry, forgot about the ResultSet API! 👀

With Async Functions (async/await) now in Node.js, this API is much easier to use! Here's an example:

```javascript
const oracledb = require('oracledb');
const config = require('./dbConfig.js');

async function processEmployees() {
  try {
    const conn = await oracledb.getConnection(config);
    const result = await conn.execute(
      'select * from employees order by employee_id',
      [],
      {resultSet: true}
    );

    let row;

    while (row  = await result.resultSet.getRow()) {
      console.log(row);
    }

    await conn.release();
  } catch (err) {
    if (conn) {
      try {
        await conn.close();
      } catch (err) {
        console.log('error closing conn', err);
      }
    }
  }
} 

processEmployees();
```

**Comment by cjbj** [comment]
@dmcghan that fetches data but concludes with:

```
(node:2780) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): ReferenceError: conn is not defined
(node:2780) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by dmcghan** [comment]
Oppps!

Here's a new and improved, albeit more verbose, version:

```javascript
const oracledb = require('oracledb');
const config = require('./dbConfig.js');

async function processEmployees() {
  let conn;
  let result;

  try {
    conn = await oracledb.getConnection(config);

    result = await conn.execute(
      'select * from employees order by employee_id',
      [],
      {resultSet: true}
    );

    let row;

    while (row  = await result.resultSet.getRow()) {
      console.log(row);
    }
  } catch (err) {
    // Will reject the promise that processEmployees returns
    // but only after finally runs
    throw err; 
  } finally {
    if (result && result.resultSet) {
      try {
        await result.resultSet.close();
      } catch (err) {
        console.log('error closing resultSet', err);
      }
    }

    if (conn) {
      try {
        await conn.close();
      } catch (err) {
        console.log('error closing conn', err);
      }
    }
  }
} 

processEmployees()
  .then(() => {
    console.log('All good!');
  })
  .catch(err => {
    console.log('Something broke!', err);
  });
```

**Comment by cjbj** [comment]
Thanks @dmcghan.  

One side note for people copying this: you will likely find it faster to use `getRows()` instead of `getRow()`.

**Comment by a2441918** [comment]
@dmcghan How do we handle 0 rows in resultSet?

                                function (err, rows) {
                                    if (err) {
                                        console.error(err);
                                        doClose(connection, resultSet); // always close the result set
                                    } else if **(rows.length > 0**) {
        
If I use "rows.length >= 0" or "rows.length > -1", I get this.

![err1](https://user-images.githubusercontent.com/4927454/32279212-6ead0bba-beee-11e7-9e87-5d5a55d9ceb2.PNG)

**Comment by dmcghan** [comment]
@a2441918 The async/await example I put above would handle this automatically. 

If you're using callbacks, the [solution in the documentation](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#-912-result-set-handling) should handle this, no?

"Callback was already called." could be related to another logical error. If you post your entire function, preferably in a new issue, we can help you debug it.

**Comment by cjbj** [comment]
Closing - no feedback.  

Check out https://blogs.oracle.com/opal/node-oracledb-v2-query-methods-and-fetch-tuning for some v2 query fetch results.
